packages/statemanager/src/ethersStateManager.ts

```diff
@@ -0,0 +1,420 @@
+    contract.storageRoot = storageTrie.root()
+    await this.putAccount(address, contract)
+  }
+
+  /**
+   * Dumps the RLP-encoded storage values for an `account` specified by `address`.
+   * @param address - The address of the `account` to return storage for
+   * @returns {Promise<StorageDump>} - The state of the account as an `Object` map.
+   * Keys are are the storage keys, values are the storage values as strings.
+   * Both are represented as `0x` prefixed hex strings.
+   */
+  dumpStorage(address: Address): Promise<StorageDump> {
+    return new Promise((resolve, reject) => {
+      this._getStorageTrie(address)
+        .then((trie) => {
+          const storage: StorageDump = {}
+          const stream = trie.createReadStream()
+
+          stream.on('data', (val: any) => {
+            storage['0x' + val.key.toString('hex')] = '0x' + val.value.toString('hex')
+          })
+          stream.on('end', () => {
+            resolve(storage)
+          })
+        })
+        .catch((e) => {
+          reject(e)
+        })
+    })
+  }
+
+  /**
+   * Checks if an `account` exists at `address`
+   * @param address - Address of the `account` to check
+   */
+  async accountExists(address: Address): Promise<boolean> {
+    log(`Verify if ${address.toString()} exists`)
+    const account = this._cache.get(address)
+    if (account.isEmpty()) {
+      // Get latest block (or block specified in `this.blockTag`)
+      const block = await this.provider.send('eth_getBlockByNumber', [this.blockTag, false])
+
+      // Get merkle proof for `address` from provider
+      const proof = await this.provider.send('eth_getProof', [address.toString(), [], block.number])
+
+      const proofBuf = proof.accountProof.map((proofNode: string) => toBuffer(proofNode))
+
+      const trie = new Trie({ useKeyHashing: true })
+      const verified = await trie.verifyProof(toBuffer(block.stateRoot), address.buf, proofBuf)
+      // if not verified (i.e. verifyProof returns null), account does not exist
+      if (verified === null) return false
+    }
+    return true
+  }
+
+  /**
+   * Gets the code corresponding to the provided `address`.
+   * @param address - Address to get the `code` for
+   * @returns {Promise<Buffer>} - Resolves with the code corresponding to the provided address.
+   * Returns an empty `Buffer` if the account has no associated code.
+   */
+  async getAccount(address: Address): Promise<Account> {
+    const account = await this._cache.getOrLoad(address)
+    return account
+  }
+
+  /**
+   * Retrieves an account from the provider and stores in the local trie
+   * @param address Address of account to be retrieved from provider
+   * @private
+   */
+  async getAccountFromProvider(address: Address): Promise<Account> {
+    const accountData = await this.provider.send('eth_getProof', [
+      address.toString(),
+      [],
+      this.blockTag,
+    ])
+
+    const rawData = accountData.accountProof
```

Comment:
should the account proof be verified here?

Reply:
I don't think so.  If the account doesn't exist in the state trie, the state manager should create a virtual account locally and the response from the provider is just an empty account (0 balance, 0 nonce, empty code/storage hash) with a proof of non existence.  We don't care if it doesn't already exist.  It would still be nice to have those proof nodes in the local state trie.

Reply:
what will ethers api return if address doesn't exist in the state? and do you mean to say the rawData proof that you got will already be validated in `await this.trie.fromProof(rawData.map((e: string) => hexToBytes(e)))`?

Reply:
Infura returns the same proof object with the account balance/nonce being 0 and the proof being one of non existence